### ISSUE TITLE
Tidy existing permissions and document

### ISSUE THREAD

**Issue by stufisher** [ispyb/py-ispyb#166]
Related to https://github.com/ispyb/py-ispyb/issues/127 we should document the existing permissions, and how these should be used from developer and user perspective.

We can also discuss `all_session`, how this can be applied to beamline groupings with `<beamline/group>_admin` instead, and how `own_sessions` can be deprecated (as all users should be able to see their sessions, it's intrinsic).

I think it's better to rely on the permissions alone and not the groups, this gives a single point reference for what a user can do. Groups should just do what they say on the tin, group users and permissions together.

**Comment by antolinos** [comment]
Hi Stuart, Could you organize a meeting for next week about authentication/authorization? 

We should discuss which are the requirements for each facility and how we can build a software that is compatible. Historically, the permission granularity has been the `proposal` (any user of a proposal has the read permissions for all the sessions). This has been demonstrated not flexible enough. You are proposing here the use of `own_sessions` by default.

I was wondering about the use of BAGs. The use of BAGs has always been a problem because the permissions are not ideally set at the level of proposal or session but of the sample. In a BAG, a group of users from different groups (and potentially competing) share the same session.

We know that the strategy is to push for the use of BAGs (and something else called hubs) as it is seen as the best way to optimize beam time and reduce paperwork. In that case, it might be necessary to implement the `own_samples` permissions or similar. I am not saying that it should be done now but we might want to double check that the system is not too constrained by today's use cases and will allow us to evolve in the future if it is needed. 

I was also wondering, how can we deal with discrepancies of the `rules`/`permissions` from different facilities? I mean, would it be possible that the software support different rules like:
- A user of a proposal has permissions to `all` proposal's sessions 
- A user of a proposal has permissions to `his` proposal's sessions
- A user of a proposal has permissions to `his own samples` (assuming it is supported by DB).

**Comment by stufisher** [comment]
Yes, these are all good questions. The historical seeing the entire proposal could be problematic as BAGs grow and contain potentially competing groups (in practice i dont think we have seen any real issues to date). In SynchWeb you can see all the sample / proteins / shipments in the proposal, but only data for sessions which you are registered on via Session_has_Person. Is this the same in the Java API / EXI?

For your use cases:
1. Can be dealt with via ProposalHasPerson (not yet implemented anywhere i dont think, but the table is agreed)
2. Use SessionHasPerson as usual
3. What to do about samples. Right now we dont know who a sample belongs to (it has no owner), and we dont want to restrict shipments as it can be that a BAG organiser ships everything but may not have registered the samples (they could have been registered by a PhD / Postdoc). This requires some thought and discussion with how others might want this to work.

cc @JPHall-DLS

**Comment by antolinos** [comment]
We authenticate with proposal accounts so it makes little sense to have permissions at level of session but we need to change that. 
My feedback is that we have received complains about confidentiality on BAGs and it is a hot topic that needs to be addressed by management first. As things might change in the coming years, I'd sleep better if py-ispyb might somehow address it.

About the use cases, can `py-ispyb` be configured in a way that:
- facility `A` implements use case `1` (A user of a proposal has permissions to all proposal's sessions)
- facility `B` implements use case `2` (A user of a proposal has permissions to his proposal's sessions)
- facility `C` implements use case `3` (A user of a proposal has permissions to his own samples)

In other words, is it foreseen that the rules/permissions are flexible and configurable with certain flexibility or we should all agree about which use case we are going to implement (Example: all facilities will set the permissions at the level of the session)?

**Comment by stufisher** [comment]
The path of least resistance is to agree on a common implementation. But its possible some of this could be configurable especially given we can abstract some of the table linkages with sqlalchemy

**Comment by ichado** [comment]
Hi all,
At SOLEIL, we do not put too much authorisation rules into ISPyB simply because the original data come from the SOLEIL User Office Tools which is called SUN set. (i e. : proposals, proposers, sessions, samples, participants, local contact...). As you know we all have existing authorisation on all these data in the User Office tools. To avoid to maintain authorisation rules from different applications and limit the the risk of error we decided to let the User Office tools manage this and simply push proposals into ISPyB. But before doing that we convert each experiment session of a bag member into a simple proposal with a specific ID (let call it BagPropX) where the members of each BagPropX are the PI, the Bag Coordinator, the participants as well as all the beamline scientists. At the end of that each user logged into ISPyB sees only the proposals on which he/she is involved. As a consequence we can put, as suggested by Andy, the Umbrella authentication on top of that with less effort on the authorization on BAG proposal at least from ISPyB side.

**Comment by antolinos** [comment]
> Hi all, At SOLEIL, we do not put too much authorisation rules into ISPyB simply because the original data come from the SOLEIL User Office Tools which is called SUN set. (i e. : proposals, proposers, sessions, samples, participants, local contact...). As you know we all have existing authorisation on all these data in the User Office tools. To avoid to maintain authorisation rules from different applications and limit the the risk of error we decided to let the User Office tools manage this and simply push proposals into ISPyB. But before doing that we convert each experiment session of a bag member into a simple proposal with a specific ID (let call it BagPropX) where the members of each BagPropX are the PI, the Bag Coordinator, the participants as well as all the beamline scientists. At the end of that each user logged into ISPyB sees only the proposals on which he/she is involved. As a consequence we can put, as suggested by Andy, the Umbrella authentication on top of that with less effort on the authorization on BAG proposal at least from ISPyB side.

That is, indeed, an interesting approach. 

> we convert each experiment session of a bag member into a simple proposal with a specific ID 

If I get it right, in SUN set you have got a BAG proposal composed of multiple sessions/visits. Then, you filled the proposal table of ISPyB with the individual BAG sessions. Threfore, your mapping is:
```
1 Sun set BAG session => 1 ISPyB Proposal 
```

That explains why the current authorization rules based on proposals works for you. 

@ichado Could you make a presentation with the details in the next auth/authorization meeting?

**Comment by ichado** [comment]
Yes of course. But I'm not available for the proposed date (17th May at 14:00).
Try to find a commun day for this presentation.

**Comment by stufisher** [comment]
> Hi all, At SOLEIL, we do not put too much authorisation rules into ISPyB simply because the original data come from the SOLEIL User Office Tools which is called SUN set. (i e. : proposals, proposers, sessions, samples, participants, local contact...). As you know we all have existing authorisation on all these data in the User Office tools. To avoid to maintain authorisation rules from different applications and limit the the risk of error we decided to let the User Office tools manage this and simply push proposals into ISPyB. But before doing that we convert each experiment session of a bag member into a simple proposal with a specific ID (let call it BagPropX) where the members of each BagPropX are the PI, the Bag Coordinator, the participants as well as all the beamline scientists. At the end of that each user logged into ISPyB sees only the proposals on which he/she is involved. As a consequence we can put, as suggested by Andy, the Umbrella authentication on top of that with less effort on the authorization on BAG proposal at least from ISPyB side.

The big downside to populating ISPyB like this is that you loose any project traceability because each proposal is a single session, proteins cannot be shared between sessions. Therefore you cannot see infer sample preparation influences data quality. Considering that it can take years to get well diffracting crystals, its likely that a protein will be collected on in multiple sessions

**Comment by antolinos** [comment]
I agree that there are pros and cons but currently, at least for us, we can not warranty the traceability of the samples neither because a protein/crystal can not be attached to multiple proposals. 
Example: BAGs have a new proposal number every year. 

One way to alleviate that constraint was the CrystalUUID that is used by CRIMS but I do not think its use is very extended. There are open discussions about using persistent identifiers (like DOI) for samples that might help on that.
